### ISSUE TITLE
BUG: `scipy.special.gammaln(2.00001)` incorrect

### ISSUE THREAD

**Issue by lucascolley** [google/jax#21284]
### Description

```python
In [2]: from jax.scipy.special import gammaln as gammaln_jax
In [5]: x = jax.numpy.asarray(2.00001)
In [7]: gammaln_jax(x)
Out[7]: Array(5.722046e-06, dtype=float32, weak_type=True)
...
In [1]: from scipy.special import gammaln
In [4]: x = np.asarray(2.00001)
In [5]: gammaln(x)
Out[5]: 4.227875597648359e-06
```

https://www.wolframalpha.com/input?i=ln%28Gamma%282.00001%29%29

Found in https://github.com/scipy/scipy/pull/20085#issuecomment-2117810225.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.27
jaxlib: 0.4.23.dev20240502
numpy:  1.26.4
python: 3.12.3 | packaged by conda-forge | (main, Apr 15 2024, 18:35:20) [Clang 16.0.6 ]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Darwin', node='Lucass-MacBook-Air-4.local', release='23.4.0', version='Darwin Kernel Version 23.4.0: Fri Mar 15 00:12:41 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T8103', machine='arm64')
```

**Comment by jakevdp** [comment]
Thanks for the report! This looks like a float precision issue. Scipy uses 64-bit precision, while JAX uses 32-bit precision by default. If you enable 64-bit precision in JAX, you get the expected output:
```python
In [1]: import jax
In [2]: jax.config.update('jax_enable_x64', True)
In [3]: x = jax.numpy.asarray(2.00001)
In [4]: jax.scipy.special.gammaln(x)
Out[4]: Array(4.2278756e-06, dtype=float64, weak_type=True)
```
That said, it seems like the 32-bit computation should be able to return a more precise answer here – in the case of `gammaln`, JAX is more-or-less directly calling OpenXLA's `lgamma` function, which is implemented here: https://github.com/openxla/xla/blob/1b9e830bc32da1e75a4bab4130376accd7e61e4d/xla/client/lib/math.cc#L513

I wonder if there's a different series expansion we could use that would be more accurate for small outputs?

**Comment by lucascolley** [comment]
Interesting, I thought that we were enabling 64-bit JAX over in SciPy. Let me check.

**Comment by lucascolley** [comment]
Feel free to close this, I re-ran CI and the test passed, so perhaps a temporary blip out of 64-bit mode for some reason.

**Comment by nstarman** [comment]
Still open on jax 0.11.1 / jaxlib 0.11.1 / scipy 1.18.1. Here is the quantification this thread was missing — "32-bit is less precise than 64-bit" doesn't separate an unavoidable loss from an avoidable one, and this one is avoidable: **scipy is correctly rounded in the same precision, and XLA is not.**

Both evaluated in float32 throughout, against 50-digit `mpmath.loggamma`:

| float32 `lgamma`, x ∈ (0.01, 10] | median ULP | p99 ULP | max ULP | max abs err |
|---|---|---|---|---|
| `jax.scipy.special.gammaln` | 3.5 | 349 | 4493 | 5.7e-06 |
| `scipy.special.gammaln` | 0.25 | 0.50 | 0.50 | 4.8e-07 |

n = 4994 random points, the two roots excluded so ULP stays meaningful. scipy sits at the 0.5 ULP bound — correctly rounded at every point in the sample. XLA's Lanczos has a median of 3.5 ULP and a p99 of 349.

```python
import numpy as np, jax.numpy as jnp, jax.scipy.special as jss, scipy.special as ss
from mpmath import mp, loggamma, mpf
mp.dps = 50

rng = np.random.default_rng(0)
x = np.unique(np.float32(rng.uniform(0.01, 10.0, 5000)))
exact = np.array([float(loggamma(mpf(float(v)))) for v in x])
jv = np.asarray(jss.gammaln(jnp.asarray(x, jnp.float32)), np.float64)
sv = np.asarray(ss.gammaln(x), np.float64)

keep = np.abs(exact) > 1e-3          # drop the roots at 1 and 2
ulp = lambda a: (np.abs(a[keep] - exact[keep])
                 / np.spacing(np.abs(exact[keep]).astype(np.float32)).astype(np.float64))
for name, u in (("jax", ulp(jv)), ("scipy", ulp(sv))):
    print(f"{name:6} median {np.median(u):8.2f}  p99 {np.percentile(u, 99):8.2f}  max {u.max():8.2f}")
```

### Where it comes from

The error is proportional to the size of the *intermediates*, not to the size of the result. `Lgamma` in [`xla/hlo/builder/lib/math.cc`](https://github.com/openxla/xla/blob/main/xla/hlo/builder/lib/math.cc) finishes with

```
log_y = log_sqrt_two_pi + (z + 1/2 - t/log_t) * log_t + log(x)
```

and every constant reaches the computation through `ScalarLike(input, ...)`, so in float32 the whole Lanczos evaluation runs in float32. Reimplementing those lines by hand at x = 1 shows what the three terms do:

```
x=1.0 f32: log_sqrt2pi=+0.918938518  mid=-6.492548466  log_A=+5.573610306  sum=+4.768371582e-07
x=1.0 f64: log_sqrt2pi=+0.918938533  mid=-6.492548490  log_A=+5.573609957  sum=+0.000000000e+00
```

Terms of magnitude 6.5 cancelling to zero. `np.spacing(np.float32(6.4925))` is `4.7683716e-07` — so the reported value is exactly **1 ULP of the largest intermediate**, and that is also exactly `2**-21`. The float32 path is already doing as well as this formulation allows; the formulation is what carries a ~5e-7 absolute error floor across the whole range, independent of how small the answer is.

That floor is invisible wherever `lgamma` is large and total wherever it is small.

### The visible symptom

`lgamma` has exact zeros at x = 1 and x = 2. In float32 JAX hits one and misses the other:

```python
>>> float(jss.gammaln(jnp.float32(1.0)))
4.76837158203125e-07
>>> float(jss.gammaln(jnp.float32(2.0)))
0.0
```

scipy returns exact zero at both; float64 is exact at both. To be clear, x = 1 is not a special failure — its neighbours carry the same absolute error (x = 0.9 is off by 5.3e-7, x = 1.35 by 8.0e-7). It is conspicuous only because the true value there is zero.

This exact example was reported in #5418 and closed as ordinary floating-point imprecision. I'd gently push back on that read: 4 ULPs of 1.0 at an exact zero is not what a correctly-rounded float32 `lgamma` produces, and scipy demonstrates as much in the same width.

### Why the error floor matters more than its size

Absolute error near a root destroys all relative accuracy, and anything forming a *difference* of log-gammas inherits it as a systematic bias rather than as noise. Binomial coefficients are the ordinary case — `exp(gammaln(N+1) - gammaln(k+1) - gammaln(N-k+1))`:

```
comb(5, 0)  ->  0.9999995231628418   # exactly exp(-gammaln(1)); should be 1
comb(5, 5)  ->  0.9999995231628418
comb(0, 0)  ->  0.9999995231628418
comb(5, 2)  ->  9.999995231628418    # should be 10
```

`comb(N, 0) = comb(N, N) = 1` is an exact identity users are entitled to rely on. It holds in float64 and fails in float32, and for `comb(N, 0)` the entirety of the error is this one value.

Happy to test any candidate float32 path against the harness above.

<sub>Downstream context: https://github.com/JAXtronomy/spexial/issues/29</sub>

**Comment by jakevdp** [comment]
> scipy is correctly rounded in the same precision, and XLA is not.

Note that in general, for float32 inputs scipy does the computation at float64 precision and then downcasts to the input dtype. This makes sense on the CPU runtimes scipy targets, as 64-bit computation is essentially universally available and performant.

JAX does not do this, because it is designed to be usable on accelerators where float64 computation is either very slow, or entirely unavailable.

As such, you'll generally find that scipy float32->float32 computations are more precise than JAX float32->float32 computations. Sometimes this is fixable in JAX without adding prohibitive computational overhead; sometimes it's not.

**Comment by hawkinsp** [comment]
https://github.com/jax-ml/jax/pull/40690 is AI generated, but it may solve your problem. My main problem with it is I'm not sure how I can get it to a readable state.
